### ISSUE TITLE
.taw files

### ISSUE THREAD

**Issue by alxwolf** [dimaryaz/jdmtool#9]
Hi, is there a way to write .taw (Terrain / Obstacle) files?

**Comment by dimaryaz** [comment]
No - but if you post your services.xml file, I can take a look at supporting them. Run `jdmtool login` and `jdmtool refresh`, then (assuming you have Linux) it should be in `~/.local/share/jdmtool/`. Just censor your username, email, customer ID, tail number, etc.

**Comment by ELD400** [comment]
Add cmd_extract_awp_taw base on your cmd_extract_awp
  to extract bin from .awp or .taw file

Tested with : j500a-us-2413.awp, t500-amrn-10t1.taw, b500-ue-24B5.taw, bdgtn-ww-24B7.taw, sgtn-eur-24S7.taw

[main.zip](https://github.com/user-attachments/files/18459178/main.zip)

**Comment by brettdoyle** [comment]
How do you manage awp files? Some of the commands mention using the extract-awp command but I looked through the repo and I don't see that as an allowable command line argument.

**Comment by dimaryaz** [comment]
It's not finished yet; I'm still trying to figure out the details of the file format. I had it in a git branch that I mostly merged into main and deleted, but forgot that it also had extract-awp.

I'll bring back the branch - but you can use the main branch and extract the .zip that ELD400 posted. Just keep it mind that it may not work correctly. It seemed to work for the navdata database. It may work for terrain/obstacles - but jdmtool does not (yet) support writing to that data card.

**Comment by dimaryaz** [comment]
(If you manage to figure out anything new, feel free to open a PR, post the code here, etc.)

**Comment by dimaryaz** [comment]
I've pushed the code to the [extract-awp-taw](https://github.com/dimaryaz/jdmtool/tree/extract-awp-taw) branch.

@ELD400: I integrated your changes, too. The second database didn't require any special logic; your code was just trying to parse the database file itself, I think. I changed it into a loop, so now it can theoretically extract more than two databases.

I changed the command so it now takes a directory as output, e.g.:

```
$ jdmtool extract-awp-taw bdgtn-ww-24B7.taw .
[a bunch of debug output]

Extracted 2 database(s):
  bdgtn-ww-24B7-0.bin
  bdgtn-ww-24B7-1.bin
```

**Comment by ELD400** [comment]
ok, I have never seen more than 2 bases per source file.

**Comment by brettdoyle** [comment]
Thanks for putting that branch in, I will give it a try. It looks like the AviaDBM has a Riptaw tool that might be useful for comparison.

**Comment by dimaryaz** [comment]
Oh nice, the rip_taw tool made it so much easier! (I'm also wondering if ELD400 was using it? The `Rgn`, `SectStart`, and `DataStart` look very familiar :) )

Anyways, I've updated the branch with lots of fixes, and even made it use the same file names as rip_taw. Though the extra "G" prefixes seem silly - GG500? Still not sure about that.

@ELD400: Can you post the file name of your IFR/VFR Charts .taw file? (I'm hoping I can download it and try it.)

**Comment by ELD400** [comment]
I try with this files:  cg1000-us-2501.taw and this b2gx000-ue-24B7.taw, dg1000-ww-24D7.taw, j2g1nx-ww-2501.taw, j2g1nx-ww-2501.taw, staxip-us-24S7.taw, tgx000-ww5-20t1.taw, fchartsp-us-2501.taw

**Comment by ELD400** [comment]
> Thanks for putting that branch in, I will give it a try. I have a Garmin 430w that I can try to load it on. It looks like the AviaDBM has a Riptaw tool that might be useful for comparison.

For information AviaDBM never support Panel ifr grm device ! (except riptaw tool)

**Comment by ELD400** [comment]
I write my first program to extract rgn in 2013 ....

Add some TAW_DATABASE_TYPES and TAW_REGION_TYPES with exact name of file 
if file go to a subfolder, I add "-" between folder 
Ex : .System-AVTN-FliteCharts-fc_tpc.dat -> e:\.System\AVTN\FliteCharts\fc_tpc.dat

if the extension is not ".bin" I add text on name
Ex: fc_tpc.fca

[main.zip](https://github.com/user-attachments/files/18596103/main.zip)

**Comment by brettdoyle** [comment]
Are there any specification documents for these file types? My understanding of formats like .awp and .taw is that they are bundles of one-to-many files with headers and trailers. I'd like to look at the structure to see if I could help out.

**Comment by dimaryaz** [comment]
@ELD400 - nice, thank you.

> ('KpGrd' start region)

Do you know what "KpGrd" means? Can there be other values? I see it in all .awp and .taw files, and rip_taw.exe contains it, so I was assuming it's some random constant.

> Add some TAW_DATABASE_TYPES and TAW_REGION_TYPES with exact name of file

I think it'll be useful to have both - a description like "Aviation" and a file path like "ldr_sys/avtn_db". Also, for data cards, the file path doesn't make sense.

> Before write file to card, grm make this for some file:

I wonder if the tail of the .awp/.taw file is a signature. It looks like completely random bytes.

**Comment by dimaryaz** [comment]
@brettdoyle: I haven't been able to find anything about .awp or .taw, so it looks like it's a custom format made up by Garmin. And I highly doubt they would want to publicly document it. If anything, they probably want to keep it secret to keep people from using non-Garmin software.

If you want to look at files other than whatever you already have: they're all hosted on `https://avdb.garmin.com/`, no authentication required - just try the file names mentioned earlier. (I don't want to post the URLs, though, to not attract Garmin's attention.)

**Comment by brettdoyle** [comment]
I found this while searching through some documentation related to Garmin Watch firmware... It looks like KpGrd is just a string placeholder, potentially telling you when the header and trailer end.

"To verify that an original file needs this header-removing, open it with hexdump -C [Download FileName] | head. You should see it starting with the string KpGrd The main watch firmware has two of those KpGrd headers, but we only remove the first one."

![Image](https://github.com/user-attachments/assets/b415437c-a8e2-448a-bbc3-80b1f5d425b8)

https://www.stevenmaude.co.uk/posts/using-garmin-forerunner-watches-with-linux

**Comment by dimaryaz** [comment]
Ah, yeah, it does seem like an arbitrary string. Maybe it's the abbreviation of the name of the Garmin employee who invented this file format.

**Comment by ELD400** [comment]
> Do you know what "KpGrd" means? Can there be other values? I see it in all .awp and .taw files, and rip_taw.exe contains it, so I was assuming it's some random constant.

I don't no, I always saw this magic word



> I think it'll be useful to have both - a description like "Aviation" and a file path like "ldr_sys/avtn_db". Also, for data cards, the file path doesn't make sense.

I think you have the long description in coverage field


> I wonder if the tail of the .awp/.taw file is a signature. It looks like completely random bytes.

it's not for all file type, but I sure for flight chart (fxxx) and Terrain (xxT1)
It's important, we don't want to hit the planet

**Comment by dimaryaz** [comment]
> I think you have the long description in coverage field
> Ex: "US/Canada/Europe Obstacles"

True. I was mostly wondering what to do when we don't know the path, or where it doesn't make sense (GNS430 data cards), but whatever. I deleted the unknown ones for now.

**Comment by ELD400** [comment]
in TAW_REGION_PATHS, all names without extention must have .bin
and "rasters_rasters.xml" is "rasters/rasters.xml"

Maybe it would be good to be able to transfer several services at the same time during downloading as for transfer :
Ex : jdmtool download 7,8,9,10

**Comment by dimaryaz** [comment]
> in TAW_REGION_PATHS, all names without extention must have .bin
> and "rasters_rasters.xml" is "rasters/rasters.xml"

Fixed

> Maybe it would be good to be able to download several services at the same time as for transfer :
> Ex : jdmtool download 7,8,9,10

Yeah, good idea, done

I've also renamed the command to "extract-taw" to keep it simpler, added an "-l" option to list the databases without extracting, and changed the output a bit.

I think it's pretty much ready to be released. (It would be nice to figure out the signature at the end... but that's not going to be easy.)

**Comment by ELD400** [comment]
> where it doesn't make sense (GNS430 data cards), but whatever. I deleted the unknown ones for now.

this may have some if we do this in 3 steps
jdmtool download 0
jdmtool extract-taw .500a-us-2413.awp
jdmtool write-database nav.bin

**Comment by dimaryaz** [comment]
Merged it into main!

**Comment by brettdoyle** [comment]
Nice I will give this a try on my GNS-430W

**Comment by brettdoyle** [comment]
I got a chance to try everything out on a GNS-430W today. It looks like everything is working great. No problems converting the taw file and then writing it to memory. Then I was able to put it in the GPS and it was formatted correctly.
